net: Encapsulate asmap in NetGroupManager

jnewbery opened the pull request:
The asmap data is currently owned by addrman, but is used by both addrman and connman. #22791 made the data const and private (so that it can't be updated by other components), but it is still passed out of addrman as a reference to const, and used by `CNetAddress` to calculate the group and AS of the net address.

This RFC PR proposes to move all asmap data and logic into a new `NetGroupManager` component. This is initialized at startup, and the client components addrman and connman simply call `NetGroupManager::GetGroup(const CAddress&)` and `NetGroupManager::GetMappedAS(const CAddress&)` to get the net group and AS of an address.

jnewbery committed:
[build] Add netgroup.cpp|h

These aren't used yet.

jnewbery committed:
[init] Add netgroupman to node.context

This is constructed before addrman and connman, and destructed afterwards.

netgroupman does not currently do anything, but will have functionality added in future commits.

jnewbery committed:
[net] Move asmap into NetGroupManager

jnewbery committed:
[netgroupman] Add GetMappedAS() and GetGroup()

These currently call through to the CNetAddr methods. The logic will be moved in a future commit.

jnewbery committed:
[netgroupman] Move GetMappedAS() and GetGroup() logic to NetGroupManager

Reviewer hint: use:

`git diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space`

jnewbery committed:
[netgroupman] Move asmap checksum calculation to NetGroupManager

jnewbery committed:
[netgroupman] Remove NetGroupManager::GetAsmap()

asmap no longer needs to be exposed anywhere outside NetGroupManager.

jnewbery committed:
[net] Only use public CNetAddr functions and data in GetMappedAS() and GetGroup()

Also change parameter/variable names. This makes the next commit mostly
move-only.

sipa commented:
Concept ACK (just judging the PR description)

practicalswift commented:
Concept ACK

naumenkogs commented:
Concept ACK, light code review ACK d70b3541cbca6c4f5b5552ec5d3fbd59bc51749a

mzumsande commented on src/test/fuzz/addrman.cpp:
```diff
@@ -101,7 +108,7 @@ class CAddrManDeterministic : public CAddrMan
                     // Add to the "tried" table (if the bucket slot is free).
                     const CAddrInfo dummy{addr, source};
-                    const int bucket = dummy.GetTriedBucket(nKey, m_asmap);
+                    const int bucket = dummy.GetTriedBucket(nKey, *g_setup->m_node.netgroupman);
```
I think this should use `m_netgroupman` which may have been fuzzed prior to calling `Fill()`, instead of the empty  `g_setup->m_node.netgroupman`, see CI fuzz test fail.

jnewbery replied:
You're absolutely right, of course. Now fixed.

mzumsande reviewed (commented):
Concept ACK

jnewbery commented:
Rebased and fixed https://github.com/bitcoin/bitcoin/pull/22910#discussion_r704821769

jnewbery commented:
rerebased

amitiuttarwar commented:
concept ACK

jnewbery commented:
I'm planning to leave this as a draft PR until #22950 is merged, since I think that's more important to get done.

jnewbery commented:
rebased on master

jnewbery commented:
I've rebased this on top of #23826, since they conflict quite heavily. Leaving as draft for now.

jnewbery commented:
Rebased on master

jnewbery commented:
I've made some very minor style/comment changes to comments, and marked this as ready for review. I was concerned that commit 76339d1750 _[net] Only use public CNetAddr functions and data in GetMappedAS() and GetGroup()_ could potentially introduce some subtle behaviour changes, but I've re-reviewed and think that it's safe. Reviewers should pay attention to the change from `m_addr` to `addr_bytes = address.GetAddrBytes()` in that commit.

MarcoFalke commented:
(This doesn't compile, see CI)

jnewbery commented:
Thanks @MarcoFalke. I've rebased on latest master and resolved the silent conflict.

fjahr commented on src/netgroup.h:
```diff
@@ -5,9 +5,41 @@
  * Netgroup manager
  */
-class NetGroupManager {};
+class NetGroupManager {
```
I feel like I am missing a bit of context on the naming decision of this class. The name is pretty generic so is the goal to add more functionality not related to ASMap? If not I would have probably opted for `AsmapMan` or so. If there is more to add here in the future maybe it would be good to add some more context in the comment above. Or maybe NetGroup is already a term used in this context and I am not aware of this?

fjahr reviewed (commented):
Code looks good from the first pass, just some naming bickering that is probably not that important. I am running an asmap node with this PR and will leave it on for at least 24 hours total before finishing the review.

fjahr commented:
ACK 5d2af606cc6c4f5decd078e86d860ea2d8009b69

Took another code review pass and didn't see any issues with the node running this code.

jnewbery commented:
Thanks for the review @fjahr. I've responded to your comments. Let me know if anything is still unclear.

mzumsande reviewed (commented):
Reviewed the code, and it looks good to me - two small comments before I Ack.

jnewbery commented on src/netaddress.cpp:
```diff
@@ -811,13 +816,14 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
+    auto addr_bytes = address.GetAddrBytes();
     const size_t num_bytes = nBits / 8;
-    vchRet.insert(vchRet.end(), m_addr.begin(), m_addr.begin() + num_bytes);
+    vchRet.insert(vchRet.end(), addr_bytes.begin() + nStartByte, addr_bytes.begin() + nStartByte + num_bytes);
```
I did consider that, but went with this approach, which is equivalent to what we were doing before commit 102867c587f5f7954232fb8ed8e85cda78bb4d32 to avoid having to expose the internal `m_addr` bytes in a public method.

jnewbery commented:
Rebased and responded to @mzumsande's review comments.

jnewbery commented on src/init.cpp:
```diff
@@ -1223,8 +1225,9 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-        assert(!node.addrman);
+        // Initialize netgroup manager
+        assert(!node.netgroupman);
+        node.netgroupman = std::make_unique<NetGroupManager>();
```
I agree. I've updated the commits so this doesn't move around. Thanks!

jnewbery commented:
rebased

jnewbery commented:
Rebased

mzumsande commented on src/addrman.cpp:
```diff
@@ -945,16 +945,16 @@ std::optional<AddressPosition> AddrManImpl::FindAddressEntry_(const CAddress& ad
-                               /*multiplicity_in=*/1,
-                               /*bucket_in=*/bucket,
+        int bucket{addr_info->GetTriedBucket(nKey, m_netgroupman.GetAsmap())};
+        return AddressPosition(/*tried=*/true,
```
This is a small rebase error, reverting [#24665 ](https://github.com/bitcoin/bitcoin/pull/24665/commits/7e22d80af333b202939bcb2631082006c097bf22) (see CI fail).

mzumsande commented:
Code Review ACK 36f814c0e84d009c0e0aa26981a20ac4cf338a85

Did some light sanity testing, and didn't run into any problems.

jnewbery commented:
CI failure seems spurious. I rebased onto latest master to trigger a new CI run, but whilst I was doing that, @mzumsande ACKed https://github.com/bitcoin/bitcoin/commit/36f814c0e84d009c0e0aa26981a20ac4cf338a85, so I've reverted to that.

dergoegge commented on src/netaddress.cpp:
```diff
@@ -778,31 +780,34 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
+    } else if (address.IsInternal()) {
+        // All internal-usage addresses get their own group.
+        // Skip over the INTERNAL_IN_IPV6_PREFIX returned by CAddress::GetAddrBytes().
+        nStartByte = INTERNAL_IN_IPV6_PREFIX.size();
```
nit: i think it would also be possible to keep the prefix and set `nBits = (INTERNAL_IN_IPV6_PREFIX.size() + ADDR_INTERNAL_SIZE) * 8;`. Then we wouldn't need to reintroduce `nStartByte`, which was previously also used for other address types.

dergoegge commented on src/netgroup.cpp:
```diff
@@ -0,0 +1,111 @@
+    // ...for the last byte, push nBits and for the rest of the byte push 1's
+    if (nBits > 0) {
+        assert(num_bytes < addr_bytes.size());
+        vchRet.push_back(addr_bytes[num_bytes] | ((1 << (8 - nBits)) - 1));
```
If `nStartByte` is ever used for other address types again it might be necessary to do the following:
```suggestion
        vchRet.push_back(addr_bytes[num_bytes + nStartByte] | ((1 << (8 - nBits)) - 1));
```
See: https://github.com/bitcoin/bitcoin/blob/1017b8a960d8bca41adebdad9982511a479689ce/src/netbase.cpp#L996

dergoegge reviewed (commented):
Code review ACK 36f814c0e84d009c0e0aa26981a20ac4cf338a85

fanquake commented:
@dergoegge you might want to open a followup with any suggestions / changes.
@jnewbery fyi I've edited your comment above to remove the @. Otherwise the ACK was being picked up and included in the merge message.

fjahr commented:
post-merge code review ACK 36f814c0e84d009c0e0aa26981a20ac4cf338a85
